pwlf/autopwlf.py

```diff
@@ -0,0 +1,232 @@
+        Examples
+        --------
+        Initialize for x, y data
+        >>> my_pwlf = AutoPiecewiseLinFit(x, y)
+        Guess number of segment first
+        >>> n = my_pwlf.fit_curve_and_guess()
+        Fit model with bounds specific
+        >>> fit_1.piecewise_fit(n, use_bound=True)
+        Calculate slopes and intercepts
+        >>> fit_1.cal_slopes()
+        Evalute model
+        >>> fit_1.model_error()
+        """
+
+
+        if not all(isinstance(n, (int, float)) for n in x) or \
+           not all(isinstance(n, (int, float)) for n in y):
+            raise ValueError("input x and y should by int or float")
+
+        x, y = self._switch_to_np_array(x), self._switch_to_np_array(y)
+        self.x_data, self.y_data = x, y
+
+    
+    @staticmethod
+    def _switch_to_np_array(input_):
+        if isinstance(input_, np.ndarray) is False:
+            input_ = np.array(input_)
+        return input_
+
+    
+    @staticmethod
+    def _unpack_expr(func):
+        r"""
+        Extract piecewise model's parameters.
+        Parameters
+        ----------
+        func : callable object
+            Model function
+        
+        Returns
+        -------
+        k : int
+            Slope
+        b : int
+            Intercept
+        """
+        text = func.__doc__
+        l = re.findall(r'return .*\n', text)[0]
+        l = re.split(r' |\n', l)[1:-1]
+        if len(l) == 3:
+            if '*x' in l[0]:
+                k, b = eval(l[0].split('*')[0]), eval(l[1]+l[2])
+            elif '*' in l[2]:
+                k, b = eval(l[1]+l[2].split('*')[0]), eval(l[0])
+        elif len(l) == 4:
+            if '*x' in l[1]:
+                k, b = eval(l[0]+l[1].split('*')[0]), eval(l[2]+l[3])
+            elif '*x' in l[3]:
+                k, b = eval(l[2]+l[3].split('*')[0]), eval(l[0]+l[1])
+        return k, b
+
+    def fit_curve_and_guess(self, degree=16, lb=10, ub=10) -> int:
+        r"""
+        Guess lines number by using curve fit then extract roots.
+        Then initize bounds with roots-lb and roots+ub, which can
+        speed up convergence.
+        Parameters
+        ----------
+        degree : int
+            Polynomial degree to fit curve 
+        lb : int
+            Lower bound
+        ub : int
+            Up bound
+        Returns
+        -------
+        guess : int
+            The guess of piecewise lines number.
+        """
+
+        coeff = np.polyfit(self.x_data, self.y_data, degree)
+        f = np.poly1d(coeff)
+        f_d = np.polyder(f, 1)
+        xroot = np.roots(f_d)
+        xroot = list(filter(lambda x : x.imag == 0, xroot))
+        xroot = list(filter(lambda x : x >= self.x_data.min() and x <= self.x_data.max(), xroot))
+        self.n = len(xroot)+1
+        self.bounds = np.zeros([self.n-1, 2])
+        for i, e in enumerate(xroot):
+            self.bounds[i][0] = e.real-lb
+            self.bounds[i][1] = e.real+ub
```

Comment:
Is the purpose of `lb` and `ub` to give some 'wiggle room' or 'slack' where break points are specified?

I think you need to make `lb ==ub` and then default this to be some percentage of the total distance of `x`.

Reply:
yes, it is 'wiggle room' or 'slack' where breakpoints are specified. I agree with you to set lb==ub = some percentage of the total distance of x